tests/lib/fakeinstaller/main.go

```diff
@@ -0,0 +1,232 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/gadget/install"
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/disks"
+	"github.com/snapcore/snapd/osutil/mkfs"
+)
+
+func firstVol(volumes map[string]*gadget.Volume) *gadget.Volume {
+	for _, vol := range volumes {
+		return vol
+	}
+	return nil
+}
+
+func createPartitions(bootDevice string, volumes map[string]*gadget.Volume) error {
+	if len(volumes) != 1 {
+		return fmt.Errorf("got unexpected number of volumes %v", len(volumes))
+	}
+
+	diskLayout, err := gadget.OnDiskVolumeFromDevice(bootDevice)
+	if err != nil {
+		return fmt.Errorf("cannot read %v partitions: %v", bootDevice, err)
+	}
+	if len(diskLayout.Structure) > 0 {
+		return fmt.Errorf("cannot yet install on a disk that has partitions")
+	}
+
+	constraints := gadget.LayoutConstraints{
+		// XXX: cargo-culted
+		NonMBRStartOffset: 1 * quantity.OffsetMiB,
+		// at this point we only care about creating partitions
+		SkipResolveContent:         true,
+		SkipLayoutStructureContent: true,
+	}
+
+	// FIXME: refactor gadget/install code to not take these dirs
+	gadgetRoot := ""
+	kernelRoot := ""
+
+	// TODO: support multiple volumes, see gadget/install/install.go
+	vol := firstVol(volumes)
+	lvol, err := gadget.LayoutVolume(gadgetRoot, kernelRoot, vol, constraints)
+	if err != nil {
+		return fmt.Errorf("cannot layout volume: %v", err)
+	}
+
+	iconst := &install.Constraints{AllPartitions: true}
+	created, err := install.CreateMissingPartitions(gadgetRoot, diskLayout, lvol, iconst)
+	if err != nil {
+		return fmt.Errorf("cannot create parititons: %v", err)
+	}
+	logger.Noticef("created %v partitions", created)
+
+	return nil
+}
+
+func runMntFor(label string) string {
+	return filepath.Join(dirs.GlobalRootDir, "/run/mnt/", label)
+}
+
+func postSystemsInstallSetupStorageEncryption(details *client.SystemDetails) error {
+	// TODO: check details.StorageEncryption and call POST
+	// /systems/<seed-label> with "action":"install" and
+	// "step":"setup-storage-encryption"
+	return nil
+}
+
+func postSystemsInstallFinish(details *client.SystemDetails) error {
+	// TODO: call POST /systems/<seed-label> with
+	// "action":"install" and "step":"finish"
```

Comment:
What is the finish step expected to do?

Reply:
I clarified this a bit now in https://github.com/snapcore/snapd/pull/12128 but I'm happy to add a comment here too.

Reply:
I also started in https://github.com/snapcore/snapd/compare/master...mvo5:systes-post-install-finish?expand=1 to flesh things out a bit more